### PR TITLE
Fix old checkpoint deletion by sorting the models properly

### LINKED ISSUE #65: "keep_ckpts" not working
In the config file, the default `"keep_ckpts": 3` does not work. When training, all checkpoints will be saved. Changing the value to any other number will not work as well.

### PR DESCRIPTION
**Pull request by Lordmau5**
Should fix #65 this time around.
The reason why this didn't work before is because it sorted them based on their name or time alone, but not also based on their prefix, so the resulting array looked like "G_", "D_", "G_", "D_", ...

This addresses that by also sorting based on their prefix, resulting in a properly sorted array "G_", "G_", "G_", "D_", "D_", "D_"

This has to be done this way due to how the itertools.groupby method works: https://note.nkmk.me/en/python-itertools-groupby/

Edit: Started training a new model from scratch with these code-changes, it's clearing it correctly
image

### CODE REVIEW

**Comment by 34j** [comment]
Thanks for the PR. However, it needs to be modified as it cannot be merged if the ci (except codecov) is not passed. Also, please use path.stem instead of .name as much as possible.

**Review by 34j — CHANGES_REQUESTED** [review]

File: src/so_vits_svc_fork/utils.py [hunk 1]
```diff
@@ -451,24 +451,26 @@ def clean_checkpoints(
...
-    for k, g in models_sorted_grouped:
-        to_dels = list(g)[:-n_ckpts_to_keep]
-        for to_del in to_dels:
-            if to_del.stem.endswith("_0"):
-                continue
-            LOG.warning(f"Removing {to_del}")
-            to_del.unlink()
+    path_key = lambda p: (p.name[0],) + (time_key(p),) if sort_by_time else (p.name[0],) + (name_key(p),)
+
+    models = list(filter(lambda p: (p.is_file() and re.match(r"[GD]_\d+\.pth", p.name) and not p.name.endswith("_0.pth")), path_to_models.glob("*.pth")))
```
[the remark this answers is not in the document]
Comment (Lordmau5) [reply src/so_vits_svc_fork/utils.py:462]: Doesn't work as that will return an empty array

File: src/so_vits_svc_fork/utils.py [hunk 2]
```diff
@@ -451,24 +451,26 @@ def clean_checkpoints(
     sort_by_time      --  True -> chronologically delete ckpts
                           False -> lexicographically delete ckpts
     """
+    LOG.warning("Cleaning old checkpoints...")
     path_to_models = Path(path_to_models)
-    name_key = lambda p: int(re.match(r"._(\d+)", p.stem).group(1))
+
+    # Define sort key functions
+    name_key = lambda p: int(re.match(r"[GD]_(\d+)\.pth", p.name).group(1))
```
[the remark this answers is not in the document]
Comment (Lordmau5) [reply src/so_vits_svc_fork/utils.py:458]: Doesn't work as that will return an empty array

**Review by 34j — APPROVED** [review]

**Comment by 34j** [comment]
@allcontributors add Lordmau5 bug, code

**Pull request merged**